Add support for computing dispatch workgroup count using program slices.

MaheshRavishankar opened the pull request:
This PR changes the way the workgroup count calculation for dispatches are handled by default.

Previous approaches required to have context of what is put into a dispatch and what values are captured as workload. Managing this implicit link gets tricky. The approach in this PR assumes that all the information required for computing the number of workgroups is captured within the dispatch after the `flow.dispatch.workgroups` op is formed. In the backend, during tile and distribute a program slice is used to capture the computation that determines the number of workgroups, and is cloned into the workgroup count region.

One complication is that the signature of the `flow.dispatch.workgroups` operation is not preserved through the compilation pipeline. So in the backends it is not possible to connect the leaves of the slice back to the operands of the original `flow.dispatch.workgroups` operation that were captured as workload. To get around this, these operands are annotated within the dispatch with their position in the workload list.

Another side-effect of this change is that there is no need to instantiate `tensor.dim` operations to compute the workload. In dynamic shape cases, this leads to repeated instantiations of the same `tensor.dim` value leading to compile-time explosion even though these get CSE-ed away later.

To address the transform dialect path a new op `transform.iree.populate_workgroup_count_region_using_num_threads_slice` is s added that handles the workgroup count materialization for the transform dialect path. As a consequence
- `transform.iree.tile_to_forall_and_workgroup_count_region` is deprecated in favor of using `transform.structured.tile_to_forall`      and the newly added transform dialect operation above.
- `transform.iree.convert_conv2d_to_img2col_and_adjust_workgroup_count_region` is also deprecated.
    

Fixes #11608

Commit:
Add support for computing dispatch workgroup count using program slices.

This PR changes the way the workgroup count calculation for dispatches
are handled by default.

Previous approaches required to have context of what is put into a
dispatch and what values are captured as workload. Managing this
implicit link gets tricky. The approach in this PR assumes that all
the information required for computing the number of workgroups is
captured within the dispatch after the flow.dispatch.workgroups op is
formed. In the backend, during tile and distribute a program slice is
used to capture the computation that determines the number of
workgroups, and is cloned into the workgroup count region.

One complication is that the signature of the
`flow.dispatch.workgroups` operation is not preserved through the
compilation pipeline. So in the backends it is not possible to connect
the leaves of the slice back to the operands of the original
flow.dispatch.workgroups operation that were captured as workload. To
get around this, these operands are annotated within the dispatch with
their position in the workload list.

Another side-effect of this change is that there is no need to
instantiate tensor.dim operations to compute the workload. In dynamic
shape cases, this leads to repeated instantiations of the same
tensor.dim value leading to compile-time explosion even though these
get CSE-ed away later.

To address the transform dialect path a new op
`transform.iree.populate_workgroup_count_region_using_num_threads_slice`
is added that handles the workgroup count materialization for the
transform dialect path. As a consequence
- `transform.iree.tile_to_forall_and_workgroup_count_region` is
  deprecated in favor of using `transform.structured.tile_to_forall`
  and the newly added transform dialect operation above.
- `transform.iree.convert_conv2d_to_img2col_and_adjust_workgroup_count_region`
  is also deprecated.

Fixes #11608

qedawkins commented on compiler/src/iree/compiler/Codegen/TransformDialectStrategies/Common/Common.cpp:
```diff
@@ -193,19 +193,23 @@ mlir::iree_compiler::buildTileFuseToScfFor(ImplicitLocOpBuilder &b,
 
+  // Handle the workgroup count region.
+  b.create<IREEPopulateWorkgroupCountRegionUsingNumThreadsSliceOp>(
+      result.forallH);
```
This needs to account for the `threadDimMapping` as well. For example if I have
```
%forall_op, %tiled_op = transform.structured.tile_to_forall_op %first   num_threads [] tile_sizes [1, 16, 32](mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>])
transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall_op : (!pdl.operation) -> () 
```
which produces something like this
```
  ^bb0(%arg0: !hal.device):
    %c16200 = arith.constant 16200 : index
    %c1 = arith.constant 1 : index
    hal.return %c16200, %c1, %c1 : index, index, index
  }
...
%6 = scf.forall (%arg0, %arg1, %arg2) in (1, 1, 16200) shared_outs(%arg3 = %5) -> (tensor<1x16x518400xf16>)
...
} {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
```
which won't map the `16200` iterator to the right block id (z).

MaheshRavishankar replied:
I don't think I fully follow. Can recommend a fix based on the current implementation

qedawkins replied:
If I specify `[#gpu.block<z>, #gpu.block<y>, #gpu.block<x>]` for the block id mapping then it works fine. The problem is that the `populate_workgroup_count_region_using_num_threads_slice` op didn't seem to look at the mapping of the loop (perhaps I am missing something somewhere else regarding usage here).

qedawkins replied:
ok I see this comment
```
  // Since the workgroup count at HAL level is in x, y, z form, process the
  // workload in reverse.
```
so is the choice of mapping `[#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]` supposed to be invalid then?

MaheshRavishankar replied:
Ok I see. If you can give me a repro to test that'd be helpful. The IR produced today and what is expected. I think I know what is missing. There is no test for this apart AFAICS

qedawkins replied:
SG, will try to send a repro later today.

qedawkins replied:
@MaheshRavishankar This example illustrates the problem
```
executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
#pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>

hal.executable private @matmul_static_dispatch_0 {
  hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
  
    hal.executable.export public @elementwise_out_of_order_block_id ordinal(0) layout(#pipeline_layout) {
    // CHECK-LABEL: hal.executable.export public @elementwise_out_of_order_block_id
    // CHECK-DAG: %[[C3:.*]] = arith.constant 3 : index
    // CHECK-DAG: %[[C5:.*]] = arith.constant 5 : index
    // CHECK-DAG: %[[C8:.*]] = arith.constant 8 : index
    // CHECK: hal.return %[[C3]], %[[C5]], %[[C8]] : index, index, index
    ^bb0(%arg0: !hal.device):
      %x, %y, %z = flow.dispatch.workgroup_count_from_body_slice
      hal.return %x, %y, %z : index, index, index
    }   
  
    builtin.module {
      func.func @elementwise_out_of_order_block_id() {
        // CHECK-LABEL: func.func @elementwise_out_of_order_block_id
        // CHECK: = scf.forall (%[[IV:.*]]) in (3, 5, 8) shared_outs(%{{.*}}) -> (tensor<3x5x8xf32>) {
        // CHECK: } {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
        %c0 = arith.constant 0 : index
        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>>
        %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
        %2 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>> -> tensor<3x5x8xf32>
        %empty = tensor.empty() : tensor<3x5x8xf32>
        %3 = linalg.generic {
          indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>,
                           affine_map<(d0, d1, d2) -> (d0, d1, d2)>],
          iterator_types = ["parallel", "parallel", "parallel"]}
          ins(%2 : tensor<3x5x8xf32>) outs(%empty : tensor<3x5x8xf32>) {   
          ^bb0(%in: f32, %in_0: f32):
            %4 = math.sqrt %in : f32 
            linalg.yield %4 : f32 
          } -> tensor<3x5x8xf32>
        flow.dispatch.tensor.store %3, %1, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : tensor<3x5x8xf32> -> !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
        return
      }   
    }   
  }
}

transform.sequence failures(propagate) {
^bb1(%variant_op: !pdl.operation):
  %1 = transform.structured.match ops{["linalg.generic"]} in %variant_op : (!pdl.operation) -> !pdl.operation
  %forall_op, %tiled_op = transform.structured.tile_to_forall_op %1   num_threads [] tile_sizes [1, 1, 1](mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>])
  transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall_op : (!pdl.operation) -> ()
}
```
Essentially this currently produces
```
hal.return %c8, %c5, %c3 : index, index, index
```
when I think it should produce
```
hal.return %c3, %c5, %c8 : index, index, index
```
to map the iteration order specified in the loop
```
%4 = scf.forall (%arg0, %arg1, %arg2) in (3, 5, 8) shared_outs(%arg3 = %3) -> (tensor<3x5x8xf32>) {
} {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
```

MaheshRavishankar replied:
> @MaheshRavishankar This example illustrates the problem
> 
> ```
> executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
> #pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>
> 
> hal.executable private @matmul_static_dispatch_0 {
>   hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
>   
>     hal.executable.export public @elementwise_out_of_order_block_id ordinal(0) layout(#pipeline_layout) {
>     // CHECK-LABEL: hal.executable.export public @elementwise_out_of_order_block_id
>     // CHECK-DAG: %[[C3:.*]] = arith.constant 3 : index
>     // CHECK-DAG: %[[C5:.*]] = arith.constant 5 : index
>     // CHECK-DAG: %[[C8:.*]] = arith.constant 8 : index
>     // CHECK: hal.return %[[C3]], %[[C5]], %[[C8]] : index, index, index
>     ^bb0(%arg0: !hal.device):
>       %x, %y, %z = flow.dispatch.workgroup_count_from_body_slice
>       hal.return %x, %y, %z : index, index, index
>     }   
>   
>     builtin.module {
>       func.func @elementwise_out_of_order_block_id() {
>         // CHECK-LABEL: func.func @elementwise_out_of_order_block_id
>         // CHECK: = scf.forall (%[[IV:.*]]) in (3, 5, 8) shared_outs(%{{.*}}) -> (tensor<3x5x8xf32>) {
>         // CHECK: } {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
>         %c0 = arith.constant 0 : index
>         %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>>
>         %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
>         %2 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>> -> tensor<3x5x8xf32>
>         %empty = tensor.empty() : tensor<3x5x8xf32>
>         %3 = linalg.generic {
>           indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>,
>                            affine_map<(d0, d1, d2) -> (d0, d1, d2)>],
>           iterator_types = ["parallel", "parallel", "parallel"]}
>           ins(%2 : tensor<3x5x8xf32>) outs(%empty : tensor<3x5x8xf32>) {   
>           ^bb0(%in: f32, %in_0: f32):
>             %4 = math.sqrt %in : f32 
>             linalg.yield %4 : f32 
>           } -> tensor<3x5x8xf32>
>         flow.dispatch.tensor.store %3, %1, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : tensor<3x5x8xf32> -> !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
>         return
>       }   
>     }   
>   }
> }
> 
> transform.sequence failures(propagate) {
> ^bb1(%variant_op: !pdl.operation):
>   %1 = transform.structured.match ops{["linalg.generic"]} in %variant_op : (!pdl.operation) -> !pdl.operation
>   %forall_op, %tiled_op = transform.structured.tile_to_forall_op %1   num_threads [] tile_sizes [1, 1, 1](mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>])
>   transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall_op : (!pdl.operation) -> ()
> }
> ```
> 
> Essentially this currently produces
> 
> ```
> hal.return %c8, %c5, %c3 : index, index, index
> ```
> 
> when I think it should produce
> 
> ```
> hal.return %c3, %c5, %c8 : index, index, index
> ```
> 
> to map the iteration order specified in the loop
> 
> ```
> %4 = scf.forall (%arg0, %arg1, %arg2) in (3, 5, 8) shared_outs(%arg3 = %3) -> (tensor<3x5x8xf32>) {
> } {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
> ```

@qedawkins Thanks! There was no lit test for this in IREE. We have a definite testing hole of these paths in IREE. I fixed the issue and added a test. Please take a look.

qedawkins reviewed (changes requested):
Verified e2e on the cases for `img2col*workgroups` as long as the implicit ordering imposed by `populate_workgroup_count_region_using_num_threads_slice` is adhered to.

Commit:
More lit test fixes.

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td:
```diff
@@ -1475,29 +1475,67 @@ def FLOW_DispatchWorkgroupCountFromDagRootOp :
+}
+
+def FLOW_DispatchWorkloadOrdinalOp :
+    FLOW_PureOp<"dispatch.workload_ordinal"> {
```
nit: `flow.dispatch.workload.ordinal` (to match `flow.dispatch.workgroup.*` ops)

benvanik reviewed (approved):
besides an op naming nit not much I can follow in here with all the transform dialect stuff - the common passes simplification is nice though and will reduce the differences between the targets ❤️

Commit:
Fix asan error?

Commit:
Address mapping issue pointed to by Quinn. Add lit test.

qedawkins reviewed (approved):
The mapping part looks good, thanks :)

Commit:
Fix transform dialect failures.

Commit:
Rebase fixes

Commit:
Clang format fixes.

Commit:
Fix bazel error.

Commit:
More bazel fixes, lit test fix and comments.

Commit:
Fix failing CUDA transform dialect tests.

Commit:
More CUDA transform dialect fixes.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/BubbleUpOrdinalOps.cpp:
```diff
@@ -0,0 +1,87 @@
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::Flow::DispatchWorkloadOrdinalOp ordinalOp,
+                                PatternRewriter &rewriter) const override {
+    auto sourceCastOp = ordinalOp.getOperand().getDefiningOp<CastOpTy>();
+    if (!sourceCastOp || sourceCastOp->hasOneUse()) {
+      return failure();
+    }
```
Maybe we can use `CastOpInterface`, so it can be applied to all the ops that implements the interface? In this context, we don't need template.

And we can add a check about if the target element type is IndexType or not.

MaheshRavishankar replied:
I am trying to slow walk this... if it fails, i want it to fail explicitly so that I can make sure things are working as expected.

hanhanW commented on compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:
```diff
@@ -252,6 +257,143 @@ template void hoistStaticallyBoundAllocationsInFunc<memref::AllocOp>(
+  return success();
+}
+
+LogicalResult lowerWorkgroupCountFromBodySliceOp(
```
I think it's better to use `rewriter.notifyMatchFailure` in this method. They are more like matching failures to me. The omitOpError is more about op verification.

MaheshRavishankar replied:
I actually want this to fail if there is a problem. So thats why I am using `emitOpError`

hanhanW reviewed (approved):
It looks like we'll need a bunch of cleanup in config materialization tests.

Commit:
Fix TF Error.

Commit:
NFC renames

`flow.dispatch.workload_ordinal` -> `flow.dispatch.workload.ordinal`
`flow.dispatch.workgroup_count_from_body_slice` -> `flow.dispatch.workgroup_count_from_slice`
